File: projects/agentic-automation-framework/src/agent/planner.js
import fs from "fs";

function fallbackWorkflowFromPrompt(prompt) {
  const p = (prompt || "").toLowerCase();
  if (p.includes("saucedemo")) {
    return {
      name: "Generated: SauceDemo login",
      url: "https://www.saucedemo.com/",
      steps: [
        { action: "type", selector: "#user-name", text: "standard_user" },
        { action: "type", selector: "#password", text: "secret_sauce" },
        { action: "click", selector: "#login-button" },
        { action: "expectUrlContains", value: "inventory.html" },
        { action: "screenshot", path: "reports/generated-saucedemo.png" }
      ]
    };
  }

  return {
    name: "Generated: Generic smoke",
    url: "https://example.com",
    steps: [
      { action: "waitFor", selector: "body" },
      { action: "screenshot", path: "reports/generated-smoke.png" }
    ]
  };
}

async function openaiPlan(prompt) {
  const key = process.env.OPENAI_API_KEY;
  if (!key) return null;

  const model = process.env.OPENAI_MODEL || "gpt-4o-mini";
  const base = (process.env.OPENAI_BASE_URL || "https://api.openai.com/v1").replace(/\/$/, "");

  const schemaHint = {
    name: "string",
    url: "https://target",
    steps: [
      { action: "type|click|waitFor|expectUrlContains|screenshot", selector: "css", text: "optional", value: "optional", path: "optional" }
    ]
  };

  const body = {
    model,
    temperature: 0.1,
    messages: [
      { role: "system", content: "You generate Playwright workflow JSON only. No markdown." },
      { role: "user", content: `Create workflow JSON from this request: ${prompt}\nSchema:${JSON.stringify(schemaHint)}` }
    ]
  };

  const res = await fetch(`${base}/chat/completions`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${key}`
    },
    body: JSON.stringify(body)
  });

  if (!res.ok) return null;
  const data = await res.json();
  const text = data?.choices?.[0]?.message?.content?.trim();
  if (!text) return null;

  try {
    const parsed = JSON.parse(text);
    return parsed;
  } catch {
    return null;
  }
}

export async function generateWorkflow(prompt) {
  const ai = await openaiPlan(prompt);
  return ai || fallbackWorkflowFromPrompt(prompt);
}

export function saveWorkflow(workflow, filePath) {
  fs.writeFileSync(filePath, JSON.stringify(workflow, null, 2));
  return filePath;
}

File: projects/agentic-automation-framework-python/agentic/agent/planner.py
import json
import requests
from agentic.config import CONFIG


def _fallback(prompt: str):
    p = (prompt or "").lower()
    if "saucedemo" in p or "login" in p:
        return {
            "name": "Generated: SauceDemo login",
            "url": "https://www.saucedemo.com/",
            "steps": [
                {"action": "type", "selector": "#user-name", "text": "standard_user"},
                {"action": "type", "selector": "#password", "text": "secret_sauce"},
                {"action": "click", "selector": "#login-button"},
                {"action": "expectUrlContains", "value": "inventory.html"},
                {"action": "screenshot", "path": "reports/generated-saucedemo.png"},
            ],
        }
    return {"name": "Generated: Generic smoke", "url": "https://example.com", "steps": [{"action": "waitFor", "selector": "body"}]}


def generate_workflow(prompt: str):
    key = CONFIG["openai"]["key"]
    if not key:
        return _fallback(prompt)

    base = CONFIG["openai"]["base_url"].rstrip("/")
    model = CONFIG["openai"]["model"]
    body = {
        "model": model,
        "temperature": 0.1,
        "messages": [
            {"role": "system", "content": "Return only valid workflow JSON. No markdown."},
            {"role": "user", "content": f"Create Playwright workflow JSON for: {prompt}"},
        ],
    }
    try:
        r = requests.post(
            f"{base}/chat/completions",
            headers={"Authorization": f"Bearer {key}", "Content-Type": "application/json"},
            json=body,
            timeout=40,
        )
        if not r.ok:
            return _fallback(prompt)
        content = r.json()["choices"][0]["message"]["content"].strip()
        return json.loads(content)
    except Exception:
        return _fallback(prompt)
